File: Knapsack/372-0_1top2bottom.py
import math
w=[1,2,3]
v=[1,5,3]
# ans=-math.inf
def dfs(i,c):
    if i==len(w):
        return 0
    ans=0
    if c>=w[i]:
        ans=max(ans,dfs(i+1,c-w[i])+v[i])
        ans=max(ans,dfs(i+1,c))
    return ans

print(dfs(0,5))

File: Knapsack/372-0_1top2bottom.cpp
#include<vector>
#include<iostream>

using VEC=std::vector<int>;


int ddffs_top2bottom(VEC& w, VEC& v, int c,int i)
{
    if(i==w.size())
        return 0;
    int ans=0;
    if(c>=w[i])
    {
        ans=ddffs_top2bottom(w,v,c-w[i],i+1)+v[i];
        ans=std::max(ans,ddffs_top2bottom(w,v,c,i+1));
    }
    return ans;
}

int main()
{
    VEC v{1,5,3};
    VEC w{1,2,3};

    int res=ddffs_top2bottom(w,v,5,0);
    std::cout<<res;
    return 0;
}